07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+			if (isResourceReused)
+				commandPool->reset();
+
+			// UPDATE DESCRIPTOR SET WRITES
+			IGPUDescriptorSet::SDescriptorInfo imgInfo;
+			IGPUImageView::SCreationParams params{};
+			params.viewType = IImageView<IGPUImage>::ET_2D;
+			params.image = images[imageToProcessId];
+			params.format = images[imageToProcessId]->getCreationParameters().format;
+			params.subresourceRange.aspectMask = IImage::E_ASPECT_FLAGS::EAF_COLOR_BIT;
+			params.subresourceRange.layerCount = images[resourceIdx]->getCreationParameters().arrayLayers;
+
+			IGPUSampler::SParams samplerParams;
+			samplerParams.AnisotropicFilter = false;
+			core::smart_refctd_ptr<IGPUSampler> sampler = m_device->createSampler(samplerParams);
+
+			imgInfo.desc = m_device->createImageView(std::move(params));
+			if (!imgInfo.desc)
+				logFailAndTerminate("Couldn't create descriptor.");
+			imgInfo.info.image = { .sampler = sampler, .imageLayout = IImage::LAYOUT::GENERAL };
+
+			IGPUDescriptorSet::SDescriptorInfo bufInfo;
+			bufInfo.desc = smart_refctd_ptr(histogramBuffer);
+			bufInfo.info.buffer = { .offset = 0u, .size = histogramBuffer->getSize() };
+
+			IGPUDescriptorSet::SWriteDescriptorSet writes[2] = {
+				{.dstSet = descSets[resourceIdx].get(), .binding = 0, .arrayElement = 0, .count = 1, .info = &imgInfo },
+				{.dstSet = descSets[resourceIdx].get(), .binding = 1, .arrayElement = 0, .count = 1, .info = &bufInfo }
+			};
+			m_device->updateDescriptorSets(2, writes, 0u, nullptr);
+
+			computeQueue->startCapture();
+			cmdBuff->begin(IGPUCommandBuffer::USAGE::NONE);
+			cmdBuff->beginDebugMarker("My Compute Dispatch", core::vectorSIMDf(0, 1, 0, 1));
+			cmdBuff->bindComputePipeline(pipeline.get());
+
+			cmdBuff->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, pplnLayout.get(), 0, 1, &descSets[resourceIdx].get());
+
+			const auto imageExtent = images[imageToProcessId]->getCreationParameters().extent;
+			const uint32_t wgCntX = imageExtent.width / WorkgroupSizeX;
+			const uint32_t wgCntY = imageExtent.height / WorkgroupSizeY;
+
+			PushConstants constants;
+			constants.histogramBufferOffset = HISTOGRAM_SIZE * resourceIdx;
+
+			cmdBuff->pushConstants(pplnLayout.get(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0u, sizeof(PushConstants), &constants);
+			cmdBuff->dispatch(wgCntX, wgCntY, 1);
+
+			cmdBuff->endDebugMarker();
+			cmdBuff->end();
+
+			IQueue::SSubmitInfo submitInfo[1];
+			IQueue::SSubmitInfo::SCommandBufferInfo cmdBuffSubmitInfo[] = { {cmdBuff.get()} };
+			IQueue::SSubmitInfo::SSemaphoreInfo signalSemaphoreSubmitInfo[] = { {.semaphore = m_imagesProcessedSemaphore.get(), .value = imageToProcessId+1, .stageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS } };
+			submitInfo[0].commandBuffers = cmdBuffSubmitInfo;
+			submitInfo[0].signalSemaphores = signalSemaphoreSubmitInfo;
+
+			waitForResourceAvailability(m_histogramSavedSeamphore.get(), imageToProcessId);
```

Comment:
you actually want to wait on `m_histogramSavedSeamphore` FRAMES_IN_FLIGHT behind

Reply:
so you'd have 2 waits [GPU waits] in the submit:
- this image is loaded `m_imagesLoadedSemaphore` 
- the histogram offset you'll be writing to is no longer being used `m_histogramSavedSeamphore+1-FRAMES_IN_FLIGHT`

both wait stages would be COMPUTE

aside from that you'd be waiting on with the CPU before going through with the commandbuffer reset, so on `m_imagesProcessedSemaphore+1-FRAMES_IN_FLIGHT`